Lib/test/test_email/test_email.py

```diff
@@ -728,6 +728,36 @@ def test_nonascii_add_header_with_tspecial(self):
             "attachment; filename*=utf-8''Fu%C3%9Fballer%20%5Bfilename%5D.ppt",
             msg['Content-Disposition'])
 
+    def test_invalid_header_names(self):
+        invalid_headers = [
+            ('Invalid Header', 'contains space'),
+            ('Tab\tHeader', 'contains tab'),
+            ('Colon:Header', 'contains colon'),
+            ('', 'Empty name'),
+            (' LeadingSpace', 'starts with space'),
+            ('TrailingSpace ', 'ends with space'),
+            ('Header\x7F', 'Non-ASCII character'),
+            ('Header\x80', 'Extended ASCII'),
+        ]
+        for thispolicy in (email.policy.default, email.policy.compat32):
+            for method in ('__setitem__', 'add_header'):
+                for name, value in invalid_headers:
+                    with self.subTest(
+                            name=name,
+                            description=value,
+                            policy=thispolicy.__class__.__name__,
+                            method=method,
+                            ):
+                        message = Message(policy=thispolicy)
+                        class_method = getattr(message, method)
+                        with self.assertRaisesRegex(
+                            ValueError,
+                            '(?i)(?=.*invalid)(?=.*header)(?=.*name)'
+                            ) as cm:
+                            class_method(name, value)
+                        self.assertIn(f"{name!r}", str(cm.exception))
```

Comment:
```suggestion
        for policy in (email.policy.default, email.policy.compat32):
            for setter in (Message.__setitem__, Message.add_header):
                for name, value in invalid_headers:
                    self.do_test_invalid_header_names(policy, setter, name, value)

    def do_test_invalid_header_names(self, policy, setter, name, value):
        with self.subTest(policy=policy, setter=setter, name=name, value=value):
            message = Message(policy=policy)
            pattern = r'(?i)(?=.*invalid)(?=.*header)(?=.*name)'
            with self.assertRaisesRegex(ValueError, pattern) as cm:
                 setter(message, name, value)
            self.assertIn(f"{name!r}", str(cm.exception))
```

This could reduce the number of lines.

Reply:
I am not sure about this optimisation. I would always have a full context instead of splitting into another function that reduces lines.

Reply:
What do you think @bitdancer ?

Reply:
> I am not sure about this optimisation. I would always have a full context instead of splitting into another function that reduces lines.

It's for readability and to ease the addition of more cases (also, please do not mark the discussion as resolved until it is really (you asked for bitdancer's opinion without waiting for them to reply))